File: editoast/schemas/src/rolling_stock/etcs_brake_params.rs
use educe::Educe;
use serde::Deserialize;
use serde::Deserializer;
use serde::Serialize;
use serde::Serializer;
use utoipa::ToSchema;

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize, ToSchema, Educe)]
#[educe(Hash)]
#[serde(deny_unknown_fields)]
/// Braking parameters for ERTMS ETCS Level 2
/// Commented with their names in ETCS specification document `SUBSET-026-3 v400.pdf` from the
/// file at <https://www.era.europa.eu/system/files/2023-09/index004_-_SUBSET-026_v400.zip>
pub struct EtcsBrakeParams {
    /// A_brake_emergency: the emergency deceleration curve (values > 0 m/s²)
    pub gamma_emergency: SpeedIntervalValueCurve,
    /// A_brake_service: the full service deceleration curve (values > 0 m/s²)
    pub gamma_service: SpeedIntervalValueCurve,
    /// A_brake_normal_service: the normal service deceleration curve used to compute guidance curve (values > 0 m/s²)
    pub gamma_normal_service: SpeedIntervalValueCurve,
    /// Kdry_rst: the rolling stock deceleration correction factors for dry rails
    /// Boundaries should be the same as gammaEmergency
    /// Values (no unit) should be contained in [0, 1]
    pub k_dry: SpeedIntervalValueCurve,
    /// Kwet_rst: the rolling stock deceleration correction factors for wet rails
    /// Boundaries should be the same as gammaEmergency
    /// Values (no unit) should be contained in [0, 1]
    pub k_wet: SpeedIntervalValueCurve,
    /// Kn+: the correction acceleration factor on normal service deceleration in positive gradients
    /// Values (in m/s²) should be contained in [0, 10]
    pub k_n_pos: SpeedIntervalValueCurve,
    /// Kn-: the correction acceleration factor on normal service deceleration in negative gradients
    /// Values (in m/s²) should be contained in [0, 10]
    pub k_n_neg: SpeedIntervalValueCurve,
    /// T_traction_cut_off: time delay in s from the traction cut-off command to the moment the acceleration due to traction is zero
    #[educe(Hash(method(common::hash_float::<5,_>)))]
    pub t_traction_cut_off: f64,
    /// T_bs1: time service break in s used for SBI1 computation
    #[educe(Hash(method(common::hash_float::<5,_>)))]
    pub t_bs1: f64,
    /// T_bs2: time service break in s used for SBI2 computation
    #[educe(Hash(method(common::hash_float::<5,_>)))]
    pub t_bs2: f64,
    /// T_be: safe brake build up time in s
    #[educe(Hash(method(common::hash_float::<5,_>)))]
    pub t_be: f64,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize, ToSchema, Educe)]
#[educe(Hash)]
#[serde(deny_unknown_fields, remote = "Self")]
pub struct SpeedIntervalValueCurve {
    #[educe(Hash(method(common::hash_float_slice::<3,_>)))]
    #[schema(example = json!([8.333333, 19.444444]))]
    /// Speed in m/s (sorted ascending)
    /// External bounds are implicit to [0, rolling_stock.max_speed]
    boundaries: Vec<f64>,
    #[educe(Hash(method(common::hash_float_slice::<3,_>)))]
    #[schema(min_items = 1, minimum = 0, example = json!([0.5, 0.6, 0.5]))]
    /// Interval values, must be >= 0 (unit to be made explicit at use)
    /// There must be one more value than boundaries
    values: Vec<f64>,
}

impl<'de> Deserialize<'de> for SpeedIntervalValueCurve {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let curve = SpeedIntervalValueCurve::deserialize(deserializer)?;

        if curve.boundaries.len() != curve.values.len() - 1 {
            return Err(serde::de::Error::custom(
                "curve invalid, expected one more value than boundaries.",
            ));
        }
        if curve.values.is_empty() {
            return Err(serde::de::Error::custom(
                "curve should have at least 1 value.",
            ));
        }
        if curve.values.iter().any(|&x| x < 0.0) {
            return Err(serde::de::Error::custom(
                "curve values must be equal or greater than 0.",
            ));
        };
        if curve.boundaries.iter().any(|&x| x < 0.0) {
            return Err(serde::de::Error::custom(
                "speed boundaries must be equal or greater than 0.",
            ));
        };
        if curve
            .boundaries
            .windows(2)
            .any(|window| window[0] >= window[1])
        {
            return Err(serde::de::Error::custom(
                "speed boundaries must be strictly increasing.",
            ));
        }

        Ok(curve)
    }
}

impl Serialize for SpeedIntervalValueCurve {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        SpeedIntervalValueCurve::serialize(self, serializer)
    }
}

File: core/envelope-sim/src/main/kotlin/fr/sncf/osrd/envelope_sim/etcs/EtcsBrakeParams.kt
package fr.sncf.osrd.envelope_sim.etcs

import fr.sncf.osrd.railjson.schema.rollingstock.RJSEtcsBrakeParams
import kotlin.math.abs

/**
 * Braking parameters for ERTMS ETCS Level 2 Commented with their names in ETCS specification
 * document `SUBSET-026-3 v400.pdf` from the file at
 * https://www.era.europa.eu/system/files/2023-09/index004_-_SUBSET-026_v400.zip
 */
class EtcsBrakeParams(
    /** A_brake_emergency: the emergency deceleration curve (values > 0 m/s²) */
    private val gammaEmergency: SpeedIntervalValueCurve,
    /** A_brake_service: the full service deceleration curve (values > 0 m/s²) */
    private val gammaService: SpeedIntervalValueCurve,
    /**
     * A_brake_normal_service: the normal service deceleration curve used to compute guidance curve
     * (values > 0 m/s²)
     */
    private val gammaNormalService: SpeedIntervalValueCurve,
    /**
     * Kdry_rst: the rolling stock deceleration correction factors for dry rails Boundaries should
     * be the same as gammaEmergency Values (no unit) should be contained in [0, 1]
     */
    private val kDry: SpeedIntervalValueCurve,
    /**
     * Kwet_rst: the rolling stock deceleration correction factors for wet rails Boundaries should
     * be the same as gammaEmergency Values (no unit) should be contained in [0, 1]
     */
    private val kWet: SpeedIntervalValueCurve,
    /**
     * Kn+: the correction acceleration factor on normal service deceleration in positive gradients
     * Values (in m/s²) should be contained in [0, 10]
     */
    private val kNPos: SpeedIntervalValueCurve,
    /**
     * Kn-: the correction acceleration factor on normal service deceleration in negative gradients
     * Values (in m/s²) should be contained in [0, 10]
     */
    private val kNNeg: SpeedIntervalValueCurve,
    /**
     * T_traction_cut_off: time delay in s from the traction cut-off command to the moment the
     * acceleration due to traction is zero
     */
    val tTractionCutOff: Double,
    /** T_bs1: time service break in s used for SBI1 computation */
    val tBs1: Double,
    /** T_bs2: time service break in s used for SBI2 computation */
    val tBs2: Double,
    /** T_be: safe brake build up time in s */
    val tBe: Double,
) {
    /** See Subset §3.13.6.2.1.4. */
    fun getSafeBrakingAcceleration(speed: Double): Double {
        val aBrakeEmergency = getEmergencyBrakingDeceleration(speed)
        val kDry = getRollingStockCorrectionFactorDry(speed)
        val kWet = getRollingStockCorrectionFactorWet(speed)
        return kDry * (kWet + mNvavadh * (1 - kWet)) * aBrakeEmergency
    }

    private fun getEmergencyBrakingDeceleration(speed: Double): Double {
        return gammaEmergency.getValue(speed)
    }

    /**
     * Corresponds to the correction factor of the emergency brake deceleration on dry tracks. The
     * confidence level mNvebcl is the confidence level that the corresponding deceleration can be
     * reached, but does not impact the calculation of kDry. See Subset §3.13.6.2.1.7.
     */
    private fun getRollingStockCorrectionFactorDry(speed: Double): Double {
        return kDry.getValue(speed)
    }

    /** Corresponds to the correction factor of the emergency brake deceleration on wet tracks. */
    private fun getRollingStockCorrectionFactorWet(speed: Double): Double {
        return kWet.getValue(speed)
    }

    fun getServiceBrakingAcceleration(speed: Double): Double {
        return gammaService.getValue(speed)
    }

    fun getNormalServiceBrakingAcceleration(speed: Double): Double {
        return gammaNormalService.getValue(speed)
    }

    /**
     * Gradient acceleration correction using on-board correction factors kN+ and kN-. See Subset,
     * §3.13.6.4.2 and §3.13.6.4.3.
     */
    fun getGradientAccelerationCorrection(grade: Double, speed: Double): Double {
        val k = if (grade >= 0) kNPos.getValue(speed) else kNNeg.getValue(speed)
        return -k * grade / 1000
    }

    class SpeedIntervalValueCurve(
        /**
         * Speed in m/s (sorted ascending). External bounds are implicit to
         * [0, rolling_stock.max_speed]
         */
        var boundaries: DoubleArray,
        /**
         * Interval values (unit to be made explicit at use) There must be one more value than
         * boundaries
         */
        var values: DoubleArray,
    ) {
        fun getValue(speed: Double): Double {
            assert(values.size == boundaries.size + 1)
            var index = 0
            val absSpeed = abs(speed)
            for (boundary in boundaries) {
                if (absSpeed <= boundary) {
                    return values[index]
                }
                index++
            }
            return values[index]
        }
    }
}

/** National Default Value: Available Adhesion. Found in Subset Appendix A.3.2 table. */
private const val mNvavadh = 0.0

fun RJSEtcsBrakeParams.toEtcsBrakeParams(): EtcsBrakeParams =
    EtcsBrakeParams(
        gammaEmergency = gammaEmergency.toSpeedIntervalValueCurve(),
        gammaService = gammaService.toSpeedIntervalValueCurve(),
        gammaNormalService = gammaNormalService.toSpeedIntervalValueCurve(),
        kDry = kDry.toSpeedIntervalValueCurve(),
        kWet = kWet.toSpeedIntervalValueCurve(),
        kNPos = kNPos.toSpeedIntervalValueCurve(),
        kNNeg = kNNeg.toSpeedIntervalValueCurve(),
        tTractionCutOff = tTractionCutOff,
        tBs1 = tBs1,
        tBs2 = tBs2,
        tBe = tBe,
    )

fun RJSEtcsBrakeParams.RJSSpeedIntervalValueCurve.toSpeedIntervalValueCurve():
    EtcsBrakeParams.SpeedIntervalValueCurve =
    EtcsBrakeParams.SpeedIntervalValueCurve(boundaries = boundaries, values = values)
